lib/text_diff.ml

```diff
@@ -119,24 +126,161 @@ let process_atomic_diff (diff : (_, _ Diff.atomic_modification) Diff.t) name
-      Same ("type " ^ type_diff.tname ^ " = {")
-      :: process_modified_record_type_diff ~indent_n:2 change_lst
-      @ [ Same "}" ]
-  | Diff.Modified (Variant_diff change_lst) ->
-      Same ("type " ^ type_diff.tname ^ " =")
-      :: process_modified_variant_type_diff change_lst
-  | Diff.Modified (Atomic mods) ->
-      process_atomic_diff (Diff.Modified mods) type_diff.tname td_to_lines
-  | Diff.Added td ->
-      process_atomic_diff (Diff.Added td) type_diff.tname td_to_lines
-  | Diff.Removed td ->
-      process_atomic_diff (Diff.Removed td) type_diff.tname td_to_lines
-
-and process_modified_record_type_diff ~indent_n diff =
+  | Modified { type_kind; type_privacy; type_manifest } ->
+      process_modified_type_diff type_diff.tname type_kind type_privacy
+        type_manifest
+  | Added td -> process_atomic_diff (Added td) type_diff.tname td_to_lines
+  | Removed td -> process_atomic_diff (Removed td) type_diff.tname td_to_lines
+
+and process_modified_type_diff name type_kind type_privacy type_manifest =
+  let type_header_diff =
+    process_type_header_diff name type_privacy type_manifest type_kind
+  in
+  let type_kind_diff = process_type_kind_diff type_kind in
+  order_type_diffs type_header_diff type_kind_diff
+
+and order_type_diffs type_header_diff type_kind_diff =
+  match (type_header_diff, type_kind_diff) with
+  | `Same _, `Same _ -> assert false
+  | `Same type_header, `Atomic_change type_kind_change
+  | `Same type_header, `Compound_change type_kind_change ->
+      type_header @ type_kind_change
+  | `Atomic_change type_header_change, `Same type_kind ->
+      type_header_change @ type_kind
+  | `Atomic_change type_header_change, `Atomic_change type_kind_change ->
+      List.hd type_header_change :: List.hd type_kind_change
+      :: List.tl type_header_change
+      @ List.tl type_kind_change
+  | `Atomic_change type_header_change, `Compound_change type_kind_change ->
+      type_header_change @ type_kind_change
+
+and process_type_header_diff name type_privacy type_manifest type_kind =
+  match (type_privacy, type_manifest, type_kind) with
+  | Left private_flag, Left te_opt, Right (Record_tk _)
+  | Left private_flag, Left te_opt, Right (Variant_tk _)
+  | ( Left private_flag,
+      Left te_opt,
+      Right
+        (Atomic_tk
+          ( { reference = Type_record (_, _); current = Type_record (_, _) }
+          | { reference = Type_record (_, _); current = Type_variant (_, _) }
+          | { reference = Type_record (_, _); current = Type_open }
+          | { reference = Type_variant (_, _); current = Type_record (_, _) }
+          | { reference = Type_variant (_, _); current = Type_variant (_, _) }
+          | { reference = Type_variant (_, _); current = Type_open }
+          | { reference = Type_open; current = Type_record (_, _) }
+          | { reference = Type_open; current = Type_variant (_, _) }
+          | { reference = Type_open; current = Type_open } )) ) ->
```

Comment:
Now that I see it in a short version, I think you can still make it much more readable with:

```ocaml
[...]
   Atomic_tk { reference = (CASE1 | CASE2 | ...); current = (CASE1 | CASE2 | ... }
```

Reply:
Also, patterns like `(_, _)` can be replaced simply by `_`)